### PR TITLE
[WIP] PodMarker - A labeling tool (#205)

### PR DESCRIPTION
**Pull request by fengzixu**
Signed-off-by: fengzixu <hnustphoenix@gmail.com>

### Ⅰ. Describe what this PR does

Add the proposal document for podMaker controller

### Ⅱ. Does this pull request fix one issue?

#205

### CODE REVIEW

**Comment by fengzixu** [comment]
@Fei-Guo @FillZpp @jian-he 

Hi guys. I submitted my proposal for implementing the podMaker controller. Can you help me review it? Actually I want to ensure that the maintainer approves my solution before I implement it.

**Comment by Fei-Guo** [comment]
Let me make upgrade logic clear in the proposal. 

- We may need to keep the history of DesiredLabels. 
- If DesiredLabels is changed, what the controller behavior should be?

**Comment by fengzixu** [comment]
> Let me make upgrade logic clear in the proposal.
> 
> * We may need to keep the history of DesiredLabels.
> * If DesiredLabels is changed, what the controller behavior should be?

In my opinion, We can implement the `version control` and `rollback` mechanism to handle the upgrade logic of PodMaker resource object.

We can dump the old spec of PodMaker before it was upgraded and mark the number. It seems like resourceVersion field. 

Ref: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#rolling-back-a-deployment

@Fei-Guo @furykerry

**Comment by Fei-Guo** [comment]
Adding version control sounds good to me. Please update the proposal with the comments and we are good to start.

**Comment by fengzixu** [comment]
@Fei-Guo @FillZpp Request review

**Comment by Fei-Guo** [comment]
> @Fei-Guo @FillZpp Request review

Mostly LG. I think you can start implementation. Btw, it is PodMarker not PodMaker.

**Comment by FillZpp** [comment]
Mostly LG, a few nits.

1. Maybe no `success`,'fail' these fields in status, just `markedReplicas`. 
`PodMarker` is not a disposable tool, Kubernetes controllers/operators are working in mode like finite-state machine. So as if a `PodMarker` existing, it should continuously try to keep `replicas` Pods to be marked.

2. Should we consider corner cases like labels overwrite, if someone creates multiple PodMarkers which have the same selector and try to mark labels with same key and different values?

**Comment by fengzixu** [comment]
>     1. disposable

For second point, I don't think it is the real corner case. The responsibility of PodMarker is simple, 1. Find the matched pods 2. Patches labels to them.

Users need to ensure that there are no two or more PodMarker objects which have same label selector.

@FillZpp @Fei-Guo

**Comment by Wenfeng-GAO** [comment]
As PodMarker only care about pod labels, for pods of the same workload(deploymnet, cloneset,etc) which possibly have same labels, how can we determine which pod/pods to mark labels?

What about adding a `Strategy` field which can possibly be:
- increasing/decreasing number order(for statefulset-like pods)
- sort by the latest pod creating time
- select pod by name
- ...

File: docs/concepts/podmarker/README.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+apiVersion: apps.kruise.io/v1alpha1
+kind: PodMarker
+metadata:
+  name: test-podmarker
+  namespace: default
+spec:
+  markLabels: # labels you want to patch to matched pods
+    upgdateStrategy: rollingUpdate
+    status: running
+  replicas: 2 # the number of pod you want tp patch labels to
```
[the remark this answers is not in the document]
Comment (fengzixu) [reply docs/concepts/podmarker/README.md:30]: Thanks

**Comment by fengzixu** [comment]
> As PodMarker only care about pod labels, for pods of the same workload(deploymnet, cloneset,etc) which possibly have same labels, how can we determine which pod/pods to mark labels?
> 
> What about adding a `Strategy` field which can possibly be:
> 
> * increasing/decreasing number order(for statefulset-like pods)
> * sort by the latest pod creating time
> * select pod by name
> * ...

I don't think that adding more complex strategies is a good idea. But you remind me, we need to think about a case "the number of the matched pod is more than replicas of podMarker". In that case, we need to consider the workload which created those pods. For example,  we may need to consider the order of pods in StatefulSet
@Wenfeng-GAO

**Pull request closed without merging**